### ISSUE TITLE
how to use type checked actions with redux middlewares

### ISSUE THREAD

**Issue by fabiosussetto** [gcanti/redux-tcomb#9]
Hi, thanks for your library! After I got a type-checked reducer to work, I'm now trying to get type-checked redux actions too. So far I have this:

```
import { t } from 'redux-tcomb'
import FieldsUnion from '../types/listColumnFields'

export const LIST_VIEW__ADD_COLUMN_AFTER = t.struct({
  viewId: t.Number,
  columnIndex: t.Number,
  newField: FieldsUnion
}, 'LIST_VIEW__ADD_COLUMN_AFTER')
```

In another action:

```
export function addColumnAfter (columnIndex) {
  return (dispatch, getState) => {
    const state = getState()
    const viewId = currentView(state).id
    const newField = textColumn({ id: 0 })

    dispatch(LIST_VIEW__ADD_COLUMN_AFTER({
      columnIndex,
      viewId,
      newField
    }))
    dispatch(generateMockData())
  }
}
```

The problem is that the type-checked action doesn't have a `type` field, so a couple of redux middleware I use (e.g. redux-logger) freak out. 
What am I missing?

Thanks!

**Comment by gcanti** [comment]
Hi, are you using the standard reducer technique https://github.com/gcanti/redux-tcomb#standard-reducer?

**Comment by fabiosussetto** [comment]
Nope, I'm using `getCheckedReducer`. The reason I don't want to use the standard reducer is mainly because I have different reducers listening for the same action type, so I don't see how the `patch` method would work in my case, as I should have different `patch` methods based on the reducer handling the action.
I'd just like to keep using getCheckedReducer (which works perfectly in my case) and then just dispatch "normal", type-checked actions, with a `type` attribute set to the struct type maybe?

**Comment by fabiosussetto** [comment]
maybe what I'm asking is related to https://github.com/gcanti/tcomb/issues/168? Just found that.

**Comment by gcanti** [comment]
Actually with "the standard reducer technique" I mean using the `getCheckedReducer` function, as linked above.

> with a type attribute set to the struct type maybe?

Yep, it should work, try this:

``` js

import { t } from 'redux-tcomb';

export const ADD_TODO = t.struct({
  type: t.enums.of('ADD_TODO'),
  text: t.String
}, 'ADD_TODO');

export const EDIT_TODO = t.struct({
  type: t.enums.of('EDIT_TODO'),
  id: t.Number,
  text: t.String
}, 'EDIT_TODO');

...

dispatch({ type: 'ADD_TODO', text: 'Use redux-tcomb' })
```

**Comment by gcanti** [comment]
Another option would be using the new `interface` feature in `tcomb` v3.1.0:

``` js
import { t } from 'redux-tcomb';

// should work even without specifying the type field
export const ADD_TODO = t.interface({
  text: t.String
}, 'ADD_TODO');

export const EDIT_TODO = t.interface({
  id: t.Number,
  text: t.String
}, 'EDIT_TODO');

...

dispatch({ type: 'ADD_TODO', text: 'Use redux-tcomb' })
```

If it works as expected, I can publish a new release (currently `redux-tcomb` depends on `tcomb` ^2.3.0)

**Comment by fabiosussetto** [comment]
Thanks. Is there a way to achieve the same thing but by doing something closer to what redux-tcomb docs show, like: 

```
ADD_TODO({ text: 'Blah' })
```

I was hoping to get rid of the action type constant name by just using the struct.

**Comment by fabiosussetto** [comment]
I came up with this solution which I thought would work:

```
// my actions module
const payloadTypes = {
  LIST_VIEW__ADD_COLUMN_AFTER: {
    viewId: t.Number,
    columnIndex: t.Number,
    newField: FieldsUnion
  },

  LIST_VIEW__REMOVE_COLUMN: {
    viewId: t.Number,
    colId: t.Number
  }
}

export default Object.keys(payloadTypes).reduce((res, type) => {
  res[type] = payload => t.struct({ type: t.String, payload: t.struct(payloadTypes[type]) }, type)({ type, payload })
  return res
}, {})

// Somewhere else
dispatch(actionTypes.LIST_VIEW__ADD_COLUMN_AFTER({
      columnIndex,
      viewId,
      newField
    }))
```

But redux complains with:
`Actions must be plain objects. Use custom middleware for async actions.`

**Comment by fabiosussetto** [comment]
At this point I think I could just move the action payload type-checking into a custom middleware to get around this issue, and just dispatch a POJO.

**Comment by gcanti** [comment]
What about adding the `type` field in a middleware?

```js
export const ADD_TODO = t.interface({
  text: t.String
}, 'ADD_TODO');

ADD_TODO.type = 'ADD_TODO'

...

const middleware = store => next => action => {
  return next(Object.assign({ type: action.constructor.type }, action))
}

...

ADD_TODO({ text: 'Blah' })
```

**Comment by gcanti** [comment]
@fabiosussetto once tcomb v3.2 is out (default props) I'd love to talk about a refactoring of redux-tcomb (a middleware seems the more idiomatic way to do type checking)

**Comment by fabiosussetto** [comment]
@gcanti sure, yeah I agree a middleware would probably be the way to go. Having a central place where you plug in type-checking for actions (the middleware) seems like a good approach, similarly to how you set up type-checking for a reducer (i.e. you don't do type checking when you handle the action in the reducer, you have a layer that does that for you automatically).

**Comment by gcanti** [comment]
POC:

``` js
import { createStore, applyMiddleware } from 'redux'
import t from 'tcomb' // gcanti/tcomb#168 branch
import createLogger from 'redux-logger'

function reducer(state, action) {
  switch(action.type) {
    case 'INCREMENT' :
      return state + action.delta
    case 'DECREMENT' :
      return state - action.delta
  }
  return state
}

// middleware
const createTypechecker = (Action, State) => store => next => action => {
  if (process.env.NODE_ENV !== 'production') {
    Action(action)
  }
  // avoid "Actions must be plain objects" error :(
  const result = next(Object.assign({}, action))
  if (process.env.NODE_ENV !== 'production') {
    State(store.getState())
  }
  return result
}

// helper
const createAction = (type, T) => {
  return t.struct.extend([T, { type: t.enums.of(type) }], {
    name: type,
    defaultProps: { type }
  })
}

// types
const PositiveInteger = t.refinement(t.Integer, n => n >= 0, 'PositiveInteger')

const Increment = createAction('INCREMENT', {
  delta: PositiveInteger
})

const Decrement = createAction('DECREMENT', {
  delta: PositiveInteger
})

const State = t.Integer
const Action = t.union([Increment, Decrement])

// configure
const store = createStore(
  reducer,
  0,
  applyMiddleware(
    createTypechecker(Action, State),
    createLogger()
  )
)

store.dispatch(Increment({ delta: 2 }))
store.dispatch(Increment({ delta: -2 })) // throws [tcomb] Invalid value -2 supplied to INCREMENT/delta: PositiveInteger
```

**Comment by fabiosussetto** [comment]
is this line in the middleware there to type-check the state after the action has been handled?

```
if (process.env.NODE_ENV !== 'production') {
    State(store.getState())
  }
```

**Comment by gcanti** [comment]
Yep

**Comment by fabiosussetto** [comment]
Maybe it's just me but I quite like the idea of having action type-checking completely independent from state type-checking, especially because I like to be able to introduce type checking only when and where I want to. 
In your POC, what if I just want type-checked actions and not reducers, or the other way round? 
Keeping the current reducer enhancer and adding the new type checking middleware could be clearer and give more flexibility too? Consider that afaik redux middlewares never make assumptions on reducers.

**Comment by gcanti** [comment]
> what if I just want type-checked actions and not reducers

``` js
createTypechecker(Action, t.Any)
```

> or the other way round

``` js
createTypechecker(t.Any, State)
```

We can also choose a different signature for `createTypechecker`:

``` js
type Options = {
  State?: Type,
  Action?: Type
};

createTypechecker(options: Options)
```

**Comment by fabiosussetto** [comment]
I see, still not very convinced though as I would expect a redux middleware to just deal with actions, not to validate the state in any way (that belongs to reducer logic)

Also (and maybe more importantly), keeping the existing reducer enhancer has these potential advantages imo:
- You can enhance only certain reducers (i.e. you want to type-check only a slice of your state)
- It's more explicit and you have more control on how you combine multiple enhancers. In my case I'm using an "undo/redo" enhancer for part of my state, and I have something like this:

```
return combineReducers({
    entities: undoable(getCheckedReducer(entities, State, t.Any),
  })
```

It's very nice that I don't have to worry about what the `undoable` reducer does to the shape of my state. Otherwise I would have to add `present`, `past` and `future` to my tcomb struct, as that thing adds those keys to manage the history for you.
- It's a bit more backward compatible with the existing implementation. People who just want to type-check their state don't need to change their code to upgrade to the new version.

What do you think?

**Comment by gcanti** [comment]
Good points. So we have:
- a `createTypechecker(Action, State)` middleware which type-checks the actions (...and optionally the state)
- a `getCheckedReducer(reducer, State, Action)` reducer enhancer which type-checks its state slice (...and optionally, but unlikely, the actions)

EDIT: <del>But then isn't the middleware redundant?</del> Nope, otherwise a "Actions must be plain objects" error is thrown

**Comment by fabiosussetto** [comment]
I think `getCheckedReducer` should drop the Action argument and just validate its state slice, but maybe that would be too big of a breaking change?

Why would the `createTypechecker` middleware be redundant? That's where you type-check the actions. I wouldn't validate the state there at all, not even as an option. On a side note, I'd call the middleware something like `actionTypeCheckMiddleware`.

Another question is where and how to define the types for your actions. In your POC you need to use the `createAction` helper, another approach could be instead to use normal actions (created without using tcomb) and then map the types in the middleware, maybe something like:

```
const actionCheckers = {
 'INCREMENT': t.struct(...)
}

// middleware
const createTypechecker = actionCheckers => store => next => action => {
  if (process.env.NODE_ENV !== 'production') {
        // from actionCheckers get the struct to validate this action against and validate it here
        // If no type defined for this action, just skip.
  }
  // No need to avoid "Actions must be plain objects" error, as they are POJOs.
  return next(action)
}

```

I'm not sure if this approach would work, users you would need to maintain the `actionCheckers` that defines how to validate each type (maybe there's a smarter way of doing that with tcomb using unions and the dispatch function?)

The potential pros could be:
- No need to convert actions to pojos
- You can type-check your existing actions without having to touch them. No need for the helper, you just need to create the `actionCheckers` and add the middleware, and keep all your existing action creators.

Not sure if this approach would be feasible though, maybe it deviates too much from the original idea?

**Comment by gcanti** [comment]
@fabiosussetto I like your proposal.

> Not sure if this approach would be feasible though, maybe it deviates too much from the original idea?

Let's start from scratch: the goal of the next version should be to write the most idiomatic library providing type-checking.

**General requirements**
- (1) No need to convert actions to pojos
- (2) You can type-check your existing actions without having to touch them

There will be 2 building blocks:
- a middleware
- a reducer enhancer

**Middleware**

The main goal of the middleware will be to type-check the actions:

``` js
// createCheckedMiddleware: (Action: Type) -> Function
export const createCheckedMiddleware = Action => store => next => action => {
  if (process.env.NODE_ENV !== 'production') {
    Action(action)
  }
  // Requirement (1) No need to convert actions to pojos
  return next(action)
}
```

**Reducer enhancer**

The main goal of the Reducer enhancer will be to type-check its state slice:

``` js
// createCheckedReducer: (reducer: Function, state: Type) -> Function
export const createCheckedReducer = (reducer, State) => (state, action) => {
  if (process.env.NODE_ENV !== 'production') {
    if (typeof state !== 'undefined') {
      State(state);
    }
  }
  state = reducer(state, action);
  if (process.env.NODE_ENV !== 'production') {
    State(state);
  }
  return state
}
```

The problem comes when you must define the `Action` type. It would be nice to have a simple way to define that union. Let's introduce an helper:

**The `createActionType` helper**

Requirements:
- the types should describe POJOs
- constants should be expressed only once

Definition of `Action`:

``` js
// using the interface combinator instead of struct
const Action = createActionType({
  INCREMENT: t.interface({ delta: PositiveInteger }), // no type field explicitly defined
  DECREMENT: t.interface({ delta: PositiveInteger })
})
```

Implementation:

``` js
// createActionType: (actions: {[key: string]: Type}) -> Type
export const createActionType = actions => {
  const dispatcher = {}
  const types = []
  const actionCreators = {}
  Object.keys(actions).forEach(type => {
    types.push(dispatcher[type] = t.interface.extend([actions[type], { type: t.enums.of(type) }], type))
    actionCreators[type] = x => t.mixin({ type }, x)
  })
  const Action = t.union(types, 'Action')
  Action.dispatch = x => dispatcher[x.type]
  t.mixin(Action, actionCreators)
  return Action
}
```

Usage:

``` js
// two choices...

// ...the standard one: you must provide a type field...
store.dispatch({ type: 'INCREMENT', delta: 2 }) // Requirement (2) You can type-check your existing actions without having to touch them

// ...or using the action creator automatically generated (no need for a type field)
store.dispatch(Action.INCREMENT({ delta: 2 }))
```

**Complete Example**

``` js
import { createStore, applyMiddleware } from 'redux'
import t from 'tcomb'
import { createCheckedMiddleware, createCheckedReducer, createActionType } from 'redux-tcomb'
import createLogger from 'redux-logger'

// types
const State = t.Integer
const PositiveInteger = t.refinement(t.Integer, n => n >= 0, 'PositiveInteger')
const Action = createActionType({
  INCREMENT: t.interface({ delta: PositiveInteger }),
  DECREMENT: t.interface({ delta: PositiveInteger })
})

// reducer
function reducer(state = 0, action) {
  switch(action.type) {
    case 'INCREMENT' :
      return state + action.delta
    case 'DECREMENT' :
      return state - action.delta
  }
  return state
}

// configure
const store = createStore(
  createCheckedReducer(reducer, State),
  applyMiddleware(
    createCheckedMiddleware(Action),
    createLogger()
  )
)

store.dispatch({ type: 'INCREMENT', delta: 2 })
store.dispatch(Action.INCREMENT({ delta: 2 }))
store.dispatch({ type: 'INCREMENT', delta: -2 }) // throws [tcomb] Invalid value -2 supplied to Action(INCREMENT)/delta: PositiveInteger
```

**Comment by fabiosussetto** [comment]
Very cool, yes that's pretty much what I had in mind but I wasn't sure about the most elegant way to do it with tcomb, need to check what t.interface does :)  I'll think a bit more about the last proposal and I'll let you know if I can think of other aspects, so far it looks good!  I may even give it a go in the app I'm building, just because it's a more realistic use case than let's say the Counter example.

Thanks,
Fabio.

**Comment by gcanti** [comment]
> I may even give it a go in the app I'm building, just because it's a more realistic use case than let's say the Counter example

It would be great, a real world use case worths 10 POCs

> need to check what t.interface does

`interface` has been introduced in tcomb v3.1, its main goal is to type-check POJOs

**Comment by fabiosussetto** [comment]
thanks for the invitation, I'm going to try what we discussed and I'll start putting together a PR maybe.

**Comment by gcanti** [comment]
Another POC using the [babel-plugin-tcomb](https://github.com/gcanti/babel-plugin-tcomb/issues/11#issuecomment-223047597) I'm working on, it wouldn't even require a dedicated redux-tcomb library: I'm type-checking directly the reducer (no middleware whatsoever). Bonus point: Flowtype compatible.

``` js
import { createStore, applyMiddleware } from 'redux'
import createLogger from 'redux-logger'

// Flow types are compiled to equivalent tcomb's types by the babel plugin
type ReduxInitAction = { type: '@@redux/INIT' }
type Action = ReduxInitAction |
  { type: 'INCREMENT', delta: number } |
  { type: 'DECREMENT', delta: number };

// reducer
function reducer(state: number = 0, action: Action): State {
  switch(action.type) {
    case 'INCREMENT' :
      return state + action.delta
    case 'DECREMENT' :
      return state - action.delta
  }
  return state
}

// configure
type Store = {
  dispatch: (action: Action) => any;
};

const store: Store = createStore(
  reducer,
  applyMiddleware(
    createLogger()
  )
)

// throws [tcomb] Invalid argument action (expected a Action)
store.dispatch({ type: 'INCREMEN', delta: 1 })
// Flow throws as well
```

**Comment by fabiosussetto** [comment]
babel-plugin-tcomb seems very powerful, I tried using it when it wasn't supporting Babel 6 yet so I'll start using it now :) Do you think then that using babel-plugin-tcomb can make this redux-specific library redundant, given how popular Babel is these days? Would there still be a valid reason to use redux-tcomb?

**Comment by gcanti** [comment]
> so I'll start using it now ... babel-plugin-tcomb can make this redux-specific library redundant

Maybe, but keep in mind that for now it's still a WIP.

> Would there still be a valid reason to use redux-tcomb?

Yes, type-checking actions in a middleware is valuable: in my POC above I'm only type-checking actions which hit the reducer but you can also have actions which hit only the middleware (if I'm not wrong).

**Comment by gcanti** [comment]
@fabiosussetto just pushed a WIP (`next` branch) as a base for reasoning

**Comment by fabiosussetto** [comment]
@gcanti thanks I haven't forgotten about this, just been a bit busy at work. I'll have a look asap.
